pkg/storage/iterator.go

```diff
@@ -77,6 +77,17 @@ type batchChunkIterator struct {
 
 // newBatchChunkIterator creates a new batch iterator with the given batchSize.
 func newBatchChunkIterator(ctx context.Context, chunks []*chunkenc.LazyChunk, batchSize int, matchers []*labels.Matcher, filter logql.Filter, req *logproto.QueryRequest) *batchChunkIterator {
+	// overlapping chunks (according to batchSize) will go through
+	// buildHeapIterator() multiple times. Because of that, we can't use the
+	// __name__ matcher since buildHeapIterator() will remove that label from the
+	// chunks that went through it.
```

Comment:
Is splicing out the `__name__` label necessary? It appears that `buildHeapIterator` removes that label if present, which is an idempotent action. If this does need to stay, I'd be interested to know why and a test case which fails if this is removed would be helpful.

Reply:
> Is splicing out the __name__ label necessary? It appears that buildHeapIterator removes that label if present, which is an idempotent action.
It is required to remove the `__name__` label matcher as long as `buildHeapIterator` removes that label from the chunks. Chunks will be mutated by `buildHeapIterator`, but those mutated chunks might be re-evaluated if they overlap. Those mutated chunks then will be discarded by the iterators since they no longer match the `__name__` label matcher.

> If this does need to stay, I'd be interested to know why and a test case which fails if this is removed would be helpful.
Every test case with overlapping chunks should fail if this is removed. It seems I botched the test cases, though. I'll fix it in the next commit and you should be able to test it with:
```shell
GOGC=10 go test -mod=vendor -p=4 pkg/storage/iterator_test.go pkg/storage/util_test.go pkg/storage/iterator.go
```
When L84-L89 are commented out, all `*_with_overlap` tests should fail.